icu4c/source/i18n/messageformat2_data_model.cpp

```diff
@@ -1094,6 +1094,8 @@ MFDataModel::MFDataModel(const MFDataModel::Builder& builder, UErrorCode& errorC
         Variant* variants = copyVectorToArray<Variant>(*builder.variants, errorCode);
         Expression* selectors = copyVectorToArray<Expression>(*builder.selectors, errorCode);
         if (U_FAILURE(errorCode)) {
+            delete variants;
```

Comment:
`copyVectorToArray` is doing an array new, so at a minimum the deletes here need to be array deletes.

Even better would be to use `LocalArray`

Reply:
PTAL